File: 2617-minimum-number-of-visited-cells-in-a-grid/2617-minimum-number-of-visited-cells-in-a-grid.cpp
class Solution {
public:
    int minimumVisitedCells(vector<vector<int>>& v){
        int m=v.size();
        int n=v[0].size();
        vector<vector<int>>dp(m,vector<int>(n,INT_MAX));
        dp[0][0]=1;
        priority_queue<pair<int,pair<int,int>>>q;
        q.push({-1,{0,0}});
        while(q.size()){
            // cout<<"df";
            int steps = (-1)*q.top().first;
            int x = q.top().second.first;
            int y = q.top().second.second;
            int val=v[x][y];
            q.pop();
            if(x==m-1 && y==n-1){
                break;
            }
            // cout<<x<<" "<<y<<endl;
            for(int i=1;i<=val && (x+i)<m;i++){
                if(dp[x+i][y]>(steps+1)){
                    dp[x+i][y]=steps+1;
                    q.push({-1*(steps+1),{x+i,y}});
                }
            }
            for(int i=1;i<=val && (y+i)<n;i++){
                if(dp[x][y+i]>(steps+1)){
                    dp[x][y+i]=steps+1;
                    q.push({-1*(steps+1),{x,y+i}});
                }
            }
            
        }
        if(dp[m-1][n-1]==INT_MAX){
            return -1;
        }
        else{
            return dp[m-1][n-1];
        }
        
    }
};

File: 2617-minimum-number-of-visited-cells-in-a-grid/2617-minimum-number-of-visited-cells-in-a-grid.java
class Solution {
    public int minimumVisitedCells(int[][] grid) {
        int m = grid.length, n = grid[0].length;
        int[][] dist = new int[m][n];
        for (int[] row : dist) 
            Arrays.fill(row, Integer.MAX_VALUE);
        dist[0][0] = 0;
        PriorityQueue<int[]> pq = new PriorityQueue<>((a, b) -> a[2] - b[2]);
        pq.offer(new int[] {0, 0, 0});
        
        while (!pq.isEmpty()) {
            int[] curr = pq.poll();
            int i = curr[0], j = curr[1], steps = curr[2];
            if (i == m-1 && j == n-1) 
                return steps+1;
            if (dist[i][j] < steps) 
                continue;
            for (int k = 1; k <= grid[i][j]; k++) 
            {
                int x = i + k, y = j;
                if (x < m && dist[x][y] > steps + 1) 
                {
                    dist[x][y] = steps + 1;
                    pq.offer(new int[] {x, y, steps + 1});
                }
                x = i; y = j + k;
                if (y < n && dist[x][y] > steps + 1) 
                {
                    dist[x][y] = steps + 1;
                    pq.offer(new int[] {x, y, steps + 1});
                }
            }
        }
        
        return -1;
    }
}